File: python/graph/dijkstra.py
from queue import PriorityQueue
def dijkstra(graph, start):
    """
    graph: 图的邻接表，元素为 (权重, 邻节点) 元组列表。
    start: 起始节点编号。
    返回各节点最短路径长度列表，无法到达则为 1e18。
    """
    dist = [1e18] * len(graph)
    dist[start] = 0
    q = PriorityQueue()
    q.put((0, start))
    while not q.empty():
        d, u = q.get()
        if d > dist[u]:
            continue
        for w, v in graph[u]:
            if dist[v] > d + w:
                dist[v] = d + w
                q.put((dist[v], v))
    return dist

if __name__ == "__main__":
    n, m, s = map(int, input().split())
    graph = [[] for _ in range(n + 1)]
    for _ in range(m):
        u, v, w = map(int, input().split())
        graph[u].append((w, v))

    dist = dijkstra(graph, s)
    print(*dist[1:])
    
    
        

File: cpp/graph/dijkstra.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
using pll = pair<ll, ll>;
using graph = vector<vector<pll>>;
constexpr ll INF = 1e18;

vector<ll> dijkstra(graph &g, ll s) {
    vector<ll> dist(g.size(), INF);
    priority_queue<pll, vector<pll>, greater<pll>> pq;
    dist[s] = 0;
    pq.emplace(0, s);
    while (!pq.empty()) {
        auto [d, u] = pq.top(); pq.pop();
        if (dist[u] < d) continue;
        for (auto [w, v] : g[u]) {
            if (dist[v] > dist[u] + w) {
                dist[v] = dist[u] + w;
                pq.emplace(dist[v], v);
            }
        }
    }
    return dist;
}

// luogu P4779 模板 单源最短路径（标准版）
int main() {
    cin.tie(nullptr)->sync_with_stdio(false);
    ll n, m, s;
    cin >> n >> m >> s;
    graph g(n + 1);
    for (int i = 0; i < m; ++i) {
        ll u, v, w;
        cin >> u >> v >> w;
        g[u].emplace_back(w, v);
    }
    auto dist = dijkstra(g, s);
    for (int i = 1; i <= n; ++i) {
        cout << dist[i] << " \n"[i == n];  
    }
}